### PR TITLE
Basic support for Kobo Nia - WiFi and USB storage export.

### PR DESCRIPTION
**Pull request by vnks**
Brief summary of the changes
----------------------------

Updated boot script to work on Nia (without this XCSoar 7 bricks the device and requires a full factory reset), and required kernel modules for WiFi and USB Storage export to work.

Related issues and discussions
------------------------------

### CODE REVIEW

File: kobo/rcS [hunk 1]
```diff
@@ -62,8 +69,19 @@ fi
 /bin/mknod /dev/null c 1 3
 /bin/chmod 666 /dev/null
 
-/sbin/udevd --daemon
-/sbin/udevadm trigger
+# udev doesn't create all /dev/ entries on some platforms (e.g. Nia),
```
[the remark this answers is not in the document]
Comment (vnks) [reply kobo/rcS:72]: Thanks for a quick review! This is what the supplied init script on the Nia is doing, so I suspect we won't fully understand it without knowing the hardware details. I'll pull out C++ changes into a separate commit, but for here our options are either to use this code, change it to uncompress the .tgz unconditionally if it's available, or use mknod to create the device tree if some required /dev/ entries are missing. My worry with unconditionally using .tgz is that it's different from what Kobo's init is doing, which is known to work on all platforms. 

Basically what happened is that sometime between 6.8 and 7 rcS switched to calling udev instead of manually creating /dev/ entires. On the Nia this does not work, and installing XCSoar 7 bricks it.

[the remark this answers is not in the document]
Comment (vnks) [reply kobo/rcS]: I poked around a bit more and and added documentation to what this is doing. Note that the same code is present in the original Kobo boot script, and in [TopHat](https://github.com/rdunning0823/tophat/blob/TopHat_master/kobo/rcS.tophat#L67), so it should allow XCSoar to boot on more platforms without impacting others. 

Let me know if you prefer to use one of the other approaches I outlined above, though personally I'd prefer to follow what is proven to work in existing boot scripts.

**Comment by MaxKellermann** [comment]
I still don't understand the rcS changes which you did for *all* Kobo devices, not just the Nia.

> On the Nia this does not work, and installing XCSoar 7 bricks it.

Did you attempt to debug it? What does the serial console say?

**Comment by vnks** [comment]
> I still don't understand the rcS changes which you did for _all_ Kobo devices, not just the Nia.

I added comments to the changes, happy to clarify what you think is missing. The /dev/ setup logic now matches what Kobo itself is doing, if you don't like it we can use one of the alternate approaches I proposed above, or I can special case it to just Nia. There would be a decent chance of bricking another Kobo as there must be a reason why they don't rely on udev for all platforms. 

> 
> > On the Nia this does not work, and installing XCSoar 7 bricks it.
> 
> Did you attempt to debug it? What does the serial console say?

I did debug it, /dev/ entries are missing, that's why it doesn't boot and that's why I needed to change the boot script. I don't have a serial console on my Kobo, this took repeated iterations of redirecting boot script output and pulling the SD card out to see what the problem was.

**Comment by vnks** [comment]
This is the commit that prevents booting on the Nia, FWIW -https://github.com/XCSoar/XCSoar/commit/2de78d4aa22cee2913db0031befa0f51019e3d2b . Adding /bin/mknod calls back in would also get it to boot.

**Comment by flightninja** [comment]
I realize this probably doesn't help much, but my current build of XCSoar 7.8 or 7.9 HEAD boots fine on my Kobo Clara HD and hasn't brick. I believe that both the Clara HD and Nickel are the Mark7 devices family so should be similar?

I wonder if something changed in a newer kernel / firmware update from Kobo? My current Kobo firmware and kernel are from around June 2020 so are not the latest, and I haven't updated it recently, but I believe there may be newer kernels / firmware available (https://github.com/kobolabs/Kobo-Reader/issues/103).

I may have to try update my Kobo firmware on my Clara and see what happens if I feel brave...

**Comment by vnks** [comment]
Clara HD and Nia are actually different platforms per [Wikipedia](https://en.wikipedia.org/wiki/Kobo_eReader#E_ink_devices), looks like Nia is the only one using the ULL chipset so far, but could be a firmware issue as well of course (I'm on the latest one for the Nia from February 2021). 

LK8000, which I think does run on the Nia, does not rely on udev for everything - they [create](https://github.com/LK8000/LK8000/blob/master/kobo/rcS#L159) some /dev/ entries via mknod. I suspect that would be enough for the Nia as well.

**Comment by vnks** [comment]
@MaxKellermann so what would you like to do here?

**Comment by MaxKellermann** [comment]
Did you address all of my comments yet?

**Comment by vnks** [comment]
> Did you address all of my comments yet?

What's missing? There's an updated request with [additional comments](https://github.com/XCSoar/XCSoar/pull/634#discussion_r642098303), and [proposed](https://github.com/XCSoar/XCSoar/pull/634#discussion_r641782517) [alternatives](https://github.com/XCSoar/XCSoar/pull/634#issuecomment-855408589).

File: kobo/rcS (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxKellermann) [reply kobo/rcS]: I'd prefer not to apply this ugly kludge to devices which don't need it. Keep the scope of bad code as small as possible.

[the remark this answers is not in the document]
Comment (vnks) [reply kobo/rcS]: It already works this way, with additional comments in https://github.com/XCSoar/XCSoar/pull/634/commits/31e922b8f2eb6c35dc62595db4f5f0d715100510. If you don't like the approach taken by the device manufacturer, you can either roll back  https://github.com/XCSoar/XCSoar/commit/2de78d4aa22cee2913db0031befa0f51019e3d2b, pick from multiple options I repeatedly provided in here, or continue bricking Kobo Nia (and likely later) devices with XCSoar 7.

[the remark this answers is not in the document]
Comment (MaxKellermann) [reply kobo/rcS]: > It already works this way

No, it doesn't. You changed the behavior of all devices, not just the Kobo Nia.

File: kobo/rcS [hunk 2]
```diff
@@ -62,8 +73,22 @@ fi
...
+# /dev/ entries. Otherwise we use the /dev tree backup if it
+# exist. This is required for platforms that don't appear to have
+# complete udev support and ship a device tree instead (such as Nia).
+if [ $PLATFORM == freescale ] || [ ! -e /etc/udev.tgz ]; then
+	/sbin/udevadm control --env=STARTUP=1
+	/sbin/udevadm trigger
+	/sbin/udevadm settle --timeout=2
+	/sbin/udevadm control --env=STARTUP=
+	# Create a /dev/ tree backup to restore next time around.
+	[ $PLATFORM != freescale ] && tar cpzf /etc/udev.tgz /dev &
```
Comment (MaxKellermann) [inline kobo/rcS:88]: This will leave a Zombie process around.

File: kobo/rcS (hunk 1 above)
[the remark this answers is not in the document]
Comment (vnks) [reply kobo/rcS]: The behavior is conditional on the platform. We're going in circles here, if you don't like this approach either pick one of the alternatives, as I keep asking, or make a concrete suggestion. This has stopped being productive a few iterations ago.

[the remark this answers is not in the document]
Comment (MaxKellermann) [reply kobo/rcS]: > The behavior is conditional on the platform.

Yes, but that only appears to contradict with my statement "You changed the behavior of all devices, not just the Kobo Nia".

> We're going in circles here

True, because I see code changes where I don't want any changes, but you don't seem to see those changes in your own code. If you want to support the Kobo Nia and take responsibility for that platform, do that, but don't change behavior for all the other Kobos. (Unless you want to be the responsible maintainer for all of them, which I'd very much welcome.)

[the remark this answers is not in the document]
Comment (vnks) [reply kobo/rcS]: > True, because I see code changes where I don't want any changes,

Then pick one of the options I've repeatedly suggested instead of ignoring them entirely, just like you did here. This entire discussion is you saying "I don't like it" to the approach taken by the device manufacturer, with no suggestions of your own nor any acknowledgement of multiple options I keep offering.

[the remark this answers is not in the document]
Comment (MaxKellermann) [reply kobo/rcS]: Maybe this wasn't clear enough:
> I'd prefer not to apply this ugly kludge to devices which don't need it. Keep the scope of bad code as small as possible.

This means: go ahead with bad code, but limit it to devices that really really need it; i.e. only Kobo Nia. Do not change anything for all other devices.

File: kobo/rcS [hunk 3]
```diff
@@ -45,6 +45,13 @@ if [ -f /mnt/onboard/.kobo/Kobo.tgz -o -f /mnt/onboard/.kobo/KoboRoot.tgz \
     exec /etc/init.d/rcS
 fi
 
+# Figure out our platform.
+PLATFORM=freescale
+if [ `dd if=/dev/mmcblk0 bs=512 skip=1024 count=1 | grep -c "HW CONFIG"` == 1 ]; then
```
[the remark this answers is not in the document]
Comment (pazos) [reply kobo/rcS:54]: Yup, all kobos from the original Touch have this blob at the same offset on unpartitioned internal storage.

File: kobo/rcS [hunk 4]
```diff
@@ -45,6 +45,13 @@ if [ -f /mnt/onboard/.kobo/Kobo.tgz -o -f /mnt/onboard/.kobo/KoboRoot.tgz \
...
+	CPU=`ntx_hwconfig -s -p /dev/mmcblk0 CPU`
```
[the remark this answers is not in the document]
Comment (pazos) [reply kobo/rcS:55]: `ntx_hwconfig` is a tool used to read blob values from linux userspace.

This will return a cpu identifier, such as 507,508, etc.

Some legacy versions might ship a blob older than the `ntx_hwconfig` tool and that will cause the command to throw some errors to `stderr`, so it is a good idea to redirect errors with:

```
CPU=$(ntx_hwconfig -s -p /dev/mmcblk0 CPU 2>/dev/null)

File: kobo/rcS (hunk 2 above)
Comment (pazos) [reply kobo/rcS:88]: This won't affect any supported model. The freescale platform only covers the original Kobo and the Kobo  Wifi, both models without touchscreen.

Anyways backgrounded jobs should be calling `waitpid()` on completion. No problem if you want to run this in the foreground either.

Comment (pazos) [reply kobo/rcS:88]: Now that I read this again it doesn't make much sense. The `tar cpzf ...` will never execute as the condition isFreescale and isNotFreescale cannot be true at the same time.

Comment (MaxKellermann) [reply kobo/rcS:88]: > Anyways backgrounded jobs should be calling `waitpid()` on completion.

Not quite - the parent process should be calling `waitpid()`, not the backgrounded job.
So, who's going to do that?

Comment (pazos) [reply kobo/rcS:88]: Not familiarized with this repo to tell you. If rcS runs the entire time because XCSoar doesn't daemonize then either insert a `wait` or just don't background. Otherwise the child will be reaped when the script finish its execution.

**But** as it's currently wroten that chunk of code won't execute, as the precondition doesn't match the specific condition to untar device nodes.

Comment (MaxKellermann) [reply kobo/rcS:88]: > If rcS runs the entire time

It does not. It `exec`s XCSoar's KoboMenu, replacing itself.

> because XCSoar doesn't daemonize

It does not, but it does not need to. The KoboMenu process just keeps running until you power the Kobo off. There's no point in daemonizing it.

**Comment by zabirauf** [comment]
@vnks Thanks for the change and @MaxKellermann thanks for helping review it. I recently ran into issue with Kobo Nia and would love to see this change merged, fixing the issues with Kobo Nia. 

Just wanted to bring this PR up as I’m curious what is remaining to get this in?

**Comment by MaxKellermann** [comment]
@zabirauf I posted the answer to your question already more than half a year ago.
(Closing this PR now because there wasn't any progress for that long time.)
